### DISCUSSION TITLE
Feed templates

### DISCUSSION THREAD

**Discussion by innermatrix** [withastro/roadmap#14] [Proposal]
Reboot of [RFC issue #1657](https://github.com/withastro/astro/issues/1657#issuecomment-977525856)

### Background & Motivation

As of 0.21, feeds generated by Astro are hard-coded in `src/build/rss.ts` and can't be customized. This is a problem because

* It's difficult to output a feed in a different format  (Atom vs RSS)
* It's difficult to output multiple feed formats (Atom and RSS) 
* While the author can insert additional data into each `<item>` via `customData`, they cannot insert any additional data into a `<channel>`.
* The custom data inserted into `<item>` is specified as a raw string, and therefore not subject to any of the syntactic conveniences of Astro/JSX templating

### Possible solutions

Broadly, two approaches have been considered. 

#### Approach a

(Originally proposed by me.) Allow an astro file to be rendered in two different modes: HTML output and feed output.

A page would opt into this by declaring a `getFeeds` function in astro frontmatter. The `getFeeds` function would return feed formats that the page wants to be rendered as. 

If a page opted into the multi-format output by declaring `getFeeds`, it would be called multiple times: once to generate HTML (unchanged from today), and once again for each feed format. The page would know whether it's being called to generate HTML or feed XML by means of a property set on the global `Astro` object.

In HTML output mode, the page would behave as it does today, producing an HTML fragment that is rendered to HTML by the compiler. In feed/XML output mode, it would be expected to produce an XML fragment, which would then be rendered by the compiler. Aside from the structural differences between XML and HTML, the other difference would be that the compiler would not wrap XML output in `<html><body>`. 

Example

```
---
# Page indicates it can generate an RSS feed.xml and an Atom atom.xml
export function getFeeds() {
  return [{
    rss2: 'feed.xml'
  }, {
    atom: 'atom.xml'
  }];
}
---
{Astro.format == "html"
? <div>[…]</div>
: Astro.format == "atom"
? <feed>[…]</feed>
: <rss>[…]</rss>
}
```

The main advantage of this approach is that it retains the current design of having any given page responsible for both its content and its feed. The main disadvantage is the potential compiler complexity of having one file that can output either HTML or XML.

#### Approach b

Proposed by @jonathantneal, as best I understand it

Allow any given astro file to output either HTML or XML, discriminated by its filename extension. One proposal is to have `.astro` files output HTML as they do today, and additionally allow `.xml.astro` files to generate XML. 

The main advantage of this design is compiler simplicity of knowing apriori whether a given source file will output XML or HTML. The main disadvantage of this design is that a collection page (which I think is the most common type of page with a feed) would need to share the collection enumeration logic between the HTML source and the XML source, leading either to code duplication or to the need for a third file to store the logic. For example:

* `category.ts` has logic to enumerate a category
* `[category].astro` has

```
---
import { getCategoryItems } from "./category.ts";

const items = getCategoryItems(…);

---
<div>[…]</div>
```

* `[category-atom].xml.astro` has

```
---
import { getCategoryItems } from "./category.ts";

const items = getCategoryItems(…);

---
<feed>[…]</feed>
```

* `[category-rss].xml.astro` has

```
---
import { getCategoryItems } from "./category.ts";

const items = getCategoryItems(…);

---
<rss>[…]</rss>
```

I believe from the previous discussions (in the issue, but possibly also from an RFC call that I did not attend) that there is broad consensus that this is a desirable feature and that either approach is viable, but no strong preference has emerged for one approach over the other as far as I am aware.

I am hoping this discussion will converge on one of those two approaches, at which point I am happy to write up an RFC for it.

**Comment by FredKSchott** [comment]
I think that https://github.com/withastro/rfcs/blob/main/active-rfcs/0006-support-non-html-files.md could address this, where you could have complete control of your RSS file without worrying about HTML templating. This is generally the direction that we'd like to go for non-HTML files, but maybe there could be some additional help for RSS feed creation (similar to our `rss()` method today).

I'll add a note about this to our v1.0 API Unification doc: #1

**Comment by Siilwyn** [comment]
What about using JSON feeds?
https://www.jsonfeed.org/ 👀

**Comment by eikowagenknecht** [comment]
Has this proposal gone anywhere? Is there a way to generate Atom feeds by now?

**Reply by innermatrix** [reply]
There is now built-in support for feeds, https://docs.astro.build/en/guides/rss/.

If you need XML output beyond what the built-in feed support offers, then you can output XML from JS/TS using an endpoint (which can be statically rendered at build time, just like any other content — https://docs.astro.build/en/guides/endpoints/), and your endpoint implementation can use any XML-generating package (I like https://www.npmjs.com/package/xmlbuilder2)
